providers/implementations/signature/rsa.c

```diff
@@ -1199,46 +1275,49 @@ static int rsa_set_ctx_params(void *vprsactx, const OSSL_PARAM params[])
                 }
             }
         }
-
-        prsactx->saltlen = saltlen;
     }
 
     p = OSSL_PARAM_locate_const(params, OSSL_SIGNATURE_PARAM_MGF1_DIGEST);
     if (p != NULL) {
-        char mdname[OSSL_MAX_NAME_SIZE] = "", *pmdname = mdname;
-        char mdprops[OSSL_MAX_PROPQUERY_SIZE] = "", *pmdprops = mdprops;
         const OSSL_PARAM *propsp =
             OSSL_PARAM_locate_const(params,
                                     OSSL_SIGNATURE_PARAM_MGF1_PROPERTIES);
 
-        if (!OSSL_PARAM_get_utf8_string(p, &pmdname, sizeof(mdname)))
+        pmgf1mdname = mgf1mdname;
+        if (!OSSL_PARAM_get_utf8_string(p, &pmgf1mdname, sizeof(mgf1mdname)))
             return 0;
 
-        if (propsp == NULL)
-            pmdprops = NULL;
-        else if (!OSSL_PARAM_get_utf8_string(propsp,
-                                             &pmdprops, sizeof(mdprops)))
-            return 0;
+        if (propsp != NULL) {
+            pmgf1mdprops = mgf1mdprops;
+            if (!OSSL_PARAM_get_utf8_string(propsp,
+                                            &pmgf1mdprops, sizeof(mgf1mdprops)))
+                return 0;
+        }
 
-        if (prsactx->pad_mode != RSA_PKCS1_PSS_PADDING) {
+        if (pad_mode != RSA_PKCS1_PSS_PADDING) {
             ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_MGF1_MD);
             return  0;
         }
+    }
 
-        if (rsa_pss_restricted(prsactx)) {
-            /* TODO(3.0) figure out what to do for prsactx->mgf1_md == NULL */
-            if (prsactx->mgf1_md == NULL
-                || EVP_MD_is_a(prsactx->mgf1_md, mdname))
-                return 1;
-            ERR_raise(ERR_LIB_PROV, PROV_R_DIGEST_NOT_ALLOWED);
-            return 0;
-        }
+    prsactx->saltlen = saltlen;
+    prsactx->pad_mode = pad_mode;
+
+    if (prsactx->md == NULL && pmdname == NULL
+        && pad_mode == RSA_PKCS1_PSS_PADDING)
+        pmdname = RSA_DEFAULT_DIGEST_NAME;
 
-        /* non-PSS code follows */
-        if (!rsa_setup_mgf1_md(prsactx, mdname, pmdprops))
+    if (pmgf1mdname != NULL
+        && !rsa_setup_mgf1_md(prsactx, pmgf1mdname, pmgf1mdprops))
+        return 0;
+
+    if (pmdname != NULL) {
+        if (!rsa_setup_md(prsactx, pmdname, pmdprops))
+            return 0;
+    } else {
+        if (!rsa_check_padding(prsactx, NULL, NULL, prsactx->mdnid))
```

Comment:
Did you intend for this path to always be run? It would get here if no valid params were passed for example.. As long as mdnid is setup before then it is probably ok?

Reply:
Yes, so the rsa_check_padding is always called when pad_mode changes. IMO this should be OK.